In [ ]:
import pandas as pd

data = pd.read_csv('/content/drive/MyDrive/heart.csv')

In [ ]:
data.head()

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
0   63    1   3     145   233    1        0       150     0      2.3    0   
1   37    1   2     130   250    0        1       187     0      3.5    0   
2   41    0   1     130   204    0        0       172     0      1.4    2   
3   56    1   1     120   236    0        1       178     0      0.8    2   
4   57    0   0     120   354    0        1       163     1      0.6    2   

   caa  thall  output  
0    0      1       1  
1    0      2       1  
2    0      2       1  
3    0      2       1  
4    0      2       1

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trtbps    303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalachh  303 non-null    int64  
 8   exng      303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slp       303 non-null    int64  
 11  caa       303 non-null    int64  
 12  thall     303 non-null    int64  
 13  output    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


In [ ]:
data.describe()

age         sex          cp      trtbps        chol         fbs  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean    54.366337    0.683168    0.966997  131.623762  246.264026    0.148515   
std      9.082101    0.466011    1.032052   17.538143   51.830751    0.356198   
min     29.000000    0.000000    0.000000   94.000000  126.000000    0.000000   
25%     47.500000    0.000000    0.000000  120.000000  211.000000    0.000000   
50%     55.000000    1.000000    1.000000  130.000000  240.000000    0.000000   
75%     61.000000    1.000000    2.000000  140.000000  274.500000    0.000000   
max     77.000000    1.000000    3.000000  200.000000  564.000000    1.000000   

          restecg    thalachh        exng     oldpeak         slp         caa  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean     0.528053  149.646865    0.326733    1.039604    1.399340    0.729373   
std      0.525860   22.905161    0.469794    1.161075    0.616226    1.022606   
min      0.000000   71.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000  133.500000    0.000000    0.000000    1.000000    0.000000   
50%      1.000000  153.000000    0.000000    0.800000    1.000000    0.000000   
75%      1.000000  166.000000    1.000000    1.600000    2.000000    1.000000   
max      2.000000  202.000000    1.000000    6.200000    2.000000    4.000000   

            thall      output  
count  303.000000  303.000000  
mean     2.313531    0.544554  
std      0.612277    0.498835  
min      0.000000    0.000000  
25%      2.000000    0.000000  
50%      2.000000    1.000000  
75%      3.000000    1.000000  
max      3.000000    1.000000

In [ ]:
data['output'].value_counts()

1    165
0    138
Name: output, dtype: int64

**1. Load the dataset and split it into training and testing sets (80% training, 20% testing).**

In [ ]:
feature = data.drop(['output'], axis=1)
label = data['output']

In [ ]:
from sklearn.model_selection import train_test_split

f_train, f_test, l_train, l_test = train_test_split(feature, label, random_state=10, test_size=0.2, stratify=label)

**2. Apply logistic regression to the training set and evaluate its accuracy on the testing set. Report the accuracy score.**

In [ ]:
from sklearn.linear_model import LogisticRegression

lr_model = LogisticRegression()

lr_model.fit(f_train, l_train)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [ ]:
lr_predict = lr_model.predict(f_test)

In [ ]:
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score

lr_report = classification_report(l_test, lr_predict)

print(lr_report)

              precision    recall  f1-score   support

           0       0.89      0.86      0.87        28
           1       0.88      0.91      0.90        33

    accuracy                           0.89        61
   macro avg       0.89      0.88      0.88        61
weighted avg       0.89      0.89      0.89        61



**3. Implement KNN with k=5 on the training set and compute its accuracy on the testing set. Report the accuracy score.**

In [ ]:
from sklearn.neighbors import KNeighborsClassifier

knn_model = KNeighborsClassifier(n_neighbors = 5)

knn_model.fit(f_train, l_train)

KNeighborsClassifier()

In [ ]:
knn_predict = knn_model.predict(f_test)

In [ ]:
knn_report = classification_report(l_test, knn_predict)

print(knn_report)

              precision    recall  f1-score   support

           0       0.60      0.64      0.62        28
           1       0.68      0.64      0.66        33

    accuracy                           0.64        61
   macro avg       0.64      0.64      0.64        61
weighted avg       0.64      0.64      0.64        61



**4. Build a decision tree classifier on the training set and evaluate its accuracy on the testing set. Report the accuracy score.**

In [ ]:
from sklearn.tree import DecisionTreeClassifier

dtc_model = DecisionTreeClassifier(max_depth = 5)

dtc_model.fit(f_train, l_train)

DecisionTreeClassifier(max_depth=5)

In [ ]:
dtc_predict = dtc_model.predict(f_test)

In [ ]:
dtc_report = classification_report(l_test, dtc_predict)

print(dtc_report)

              precision    recall  f1-score   support

           0       0.73      0.68      0.70        28
           1       0.74      0.79      0.76        33

    accuracy                           0.74        61
   macro avg       0.74      0.73      0.73        61
weighted avg       0.74      0.74      0.74        61



**5. Compare the accuracy scores of logistic regression, KNN, and decision trees. Which algorithm performed the best in this case?**

In [ ]:
from sklearn.metrics import accuracy_score
import numpy as np

model = {'LogisticRegression' : LogisticRegression(),
         'KNeighborsClassifier' : KNeighborsClassifier(n_neighbors = 5),
         'DecisionTreeClassifier' : DecisionTreeClassifier(max_depth = 5)}


for name, obj in model.items():
  obj.fit(f_train, l_train)
  prediction = obj.predict(f_test)

  print(f'{name} : {np.round(accuracy_score(l_test, prediction), 2)}')


LogisticRegression : 0.89
KNeighborsClassifier : 0.64
DecisionTreeClassifier : 0.74


/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


**6. Choose one of the algorithms that performed well and briefly discuss why it might be suitable for this specific problem.**

=> Based on the accuracy score, the Logistic Regression model performs exceptionally well with a high accuracy of 89%.

=> This model is particularly suitable for the task at hand because it is designed for binary classification problems, which involve classifying data into two categories (0 and 1).

=> While Logistic Regression is well-suited for binary classification, it can also handle multi-class classification problems without the need for additional parameters or modifications


**Some other reasons also which makes this model ore suitable to use -**

Interpretability: The coefficients of the model can be interpreted as the effect of each feature on the likelihood of a particular outcome. This makes it easier to understand the importance of each feature in the classification.

Efficiency: Logistic Regression is computationally efficient and works well with both small and large datasets.

Low Risk of Overfitting: Logistic Regression tends to be less prone to overfitting compared to more complex models.

Probabilistic Output: Logistic Regression models provide probabilistic outputs, which can be used to rank instances by their likelihood of belonging to a particular class.

Regularization: Techniques like L1 (Lasso) and L2 (Ridge) regularization can be applied to handle multicollinearity and reduce model complexity.

Feature Importance: Logistic Regression allows for the assessment of feature importance, helping to identify which features are most relevant in making classification decisions.

Wide Applicability: While commonly used for binary classification, Logistic Regression can also be extended to handle multi-class classification problems with techniques like one-vs-all (OvA) or softmax regression.

Simple to Implement: Implementing Logistic Regression is relatively straightforward and requires minimal hyperparameter tuning.

**7. What are some potential ways to improve the performance of the chosen algorithm?**

There are various ways to improve the Logistic Regression model performance _

=> Feature selection: Remove irrelevant or redundant features.

=> Cross-validation: Employ cross-validation techniques for better validation and results.

=> Hyperparameter tuning: Optimize model parameters for improved performance.

=> Feature engineering: Introduce important, engineered, or relevant features.

=> Data preprocessing: Prepare and clean the data for better model training.

=> Regularization: Apply regularization techniques to prevent overfitting.

=> Feature scaling/normalization: Scale features for consistent influence on the model.

=> Outlier handling: Address outliers in the data appropriately.

=> Imbalanced data: Balance imbalanced datasets for fair model training.

=> Categorical feature encoding: Convert categorical features into a suitable format.

=> Missing value imputation: Handle missing data for more robust modeling

**8. Evaluate the importance of each feature in the decision tree classifier. Plot a bar chart showing the feature importances.**



**9. Explain one practical application for each of the three algorithms: logistic regression, KNN, and decision trees.**



**Logistic Regression:**


**Spam Email Detection:** Logistic regression is used to classify emails as spam or not spam based on various features, making it a common choice for email filtering.

**Credit Card Fraud Detection:** Logistic regression helps identify fraudulent credit card transactions by modeling the probability of a transaction being fraudulent based on historical data.

**K-Nearest Neighbors (KNN):**

**Recommendation Systems:** KNN is employed in collaborative filtering to recommend products, movies, or content by finding similar users' preferences.

**Anomaly Detection in Network Security:** KNN identifies unusual patterns or intrusions in network data by comparing network activity to the behavior of normal instances.

**Decision Trees:**

**Customer Segmentation in Marketing:** Decision trees segment customers into groups with similar characteristics, aiding marketers in targeting specific demographics for tailored campaigns.

**Medical Diagnosis:** Decision trees are used to assist in medical diagnosis by considering patient symptoms and medical history to determine likely conditions or diseases.


**10. In what scenarios would you prefer to use logistic regression over KNN and decision trees, and vice versa?**

**Prefer Logistic Regression:**

=> when dealing with a binary classification problem, where the outcome is either 0 or 1.

=> when the relationship between the dependent variable and the independent variables is assumed to be linear on the log-odds scale.

=>  when you need probabilities of class membership for each observation.
When to Prefer K-Nearest Neighbors (KNN):

**Prefer KNN**

=> when there is no prior assumption about the distribution of the data, and the decision boundary can be complex and nonlinear.

=> It's useful for both classification and regression tasks.

=> KNN can be effective for small to medium-sized datasets, where calculating distances between data points is computationally feasible.

**Prefer Decision Trees:**

=> when you want to interpret the model's decision-making process, as they provide clear and interpretable rules for classification or regression.

=> when the dataset contains a mix of categorical and numerical variables.

=> when you want a model that can handle missing data and outliers effectively.




